content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md

```diff
@@ -0,0 +1,538 @@
+   return offset
+}
+// inline
+func (x *User) fastWriteField2(buf []byte) (offset int) {
+   offset += fastpb.WriteString(buf[offset:], 2, x.Name)
+   return offset
+}
+```
+
+### Optimization Results
+
+As a result of these optimizations, we managed to improve the optimization output from the previous contribution of 3.58% to a notable 0.98%.
+
+![Image](/img/blog/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/9.jpeg)
+
+## JIT Substitution for Code Generation: Introducing Frugal (Thrift)
+
+After reaping gains from the hardcoded approach, we encountered the following feedback:
+
+1. The volume of the generated code increases linearly with the growth of fields.
+
+2. The generated code depends on the user's version of their respective Kitex command-line tool, which can lead to conflicts during collaborations with multiple contributors.
+
+The feedback encouraged us to consider if the previously generated code could be created automatically at runtime. The answer is 'yes' - but it would require the adoption of Just-In-Time Compilation (JIT) as a method of code optimization.
+
+![Image](/img/blog/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/10.jpeg)
+
+### Advantages of JIT 
+
+1. Registers utilization and deeper inlining: this improves the efficiency of function calls.
+
+2. Core computational functions use fully optimized assembly code, which leads to improved performance.
+
+### Optimization Results of JIT
+As a result of JIT optimization, we improved the optimization result from 3.58% to an impressive 0.78%.
+
+![Image](/img/blog/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/11.jpeg)
+
+### Comparing Frugal and Apache Thrift 
+
+This section presents a performance comparison of Frugal and Apache Thrift in the context of encoding and decoding.
+
+![Image](/img/blog/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/12.jpeg)
+
+## Optimizing Network Library
+
+![Image](/img/blog/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/13.jpeg)
+
+### Native Go Net Challenges in RPC Scenarios
+Native Go Net in RPC situations presents the following challenges:
+
+1. Each connection corresponds to one coroutine - when there are numerous upstream and downstream instances, the sheer number of Goroutines can significantly influence performance. This is particularly detrimental for businesses with intensive instances.
+
+2. It is unable to automatically detect the connection's shutdown state.
+
+3. When a struct undergoes NoCopy serialization, the output typically takes the form of a two-dimensional byte array. However, Go's `Write([]byte)` interface falls short as it does not support handling non-continuous memory data.
+
+4. Despite being highly compatible, it's provided by Go Runtime and not conducive or suitable for adding new features. 
+
+![Image](/img/blog/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/14.jpeg)
+
+#### Go Coding Example
+
+```Go
```

Comment:
Not placed correctly according to the original doc.

Reply:
I moved and changed a few things. If the code and logic are correct, it's better to leave it like this. It's not a 1 to 1 translation